jina/flow/base.py

```diff
@@ -708,6 +708,10 @@ def add(
             kwargs, parser, True, fallback_parsers=FALLBACK_PARSERS
         )
 
+        # grpc data runtime does not support sharding at the moment
+        if kwargs.get('shards') is not None:
+            raise NotImplementedError("GRPC data runtime does not support sharding")
+
         if args.grpc_data_requests and args.runtime_cls == 'ZEDRuntime':
             args.runtime_cls = 'GRPCDataRuntime'
 
```

Comment:
Just pitching in here because I had been following this issue for a while. 

Since, we are concerned about not using GrpcDataRuntime with shards, I believe that the if that has been added in line 712 should come in line 718 after the completion of if where we are assigning runtime_cls to GRPCDataRuntime. Also, there should be a check to see if runtime_cls is GRPCDataRuntime before raising the NotImplementedError. The current implementation will not allow sharding with any runtime. For example, when args.grpc_data_requests is false and args.runtime_cls == 'ZEDRuntime', we move ahead with ZEDRuntime but are still not allowing sharding. This should not be the case.

Reply:
I think it should be independent of the actual `runtime_cls` so I would not do the check inside `if args.grpc_data_requests and args.runtime_cls == 'ZEDRuntime':`

But you are right that the current check will prevent any sharding, so the check should be something like 

if args.grpc_data_requests and kwargs.get('shards') is not None:
    raise NotImplementedError("GRPC data runtime does not support sharding")

Reply:
If we check for grpc_data_requests and shards together, then I think current placement is fine.